### PR TITLE
LoRA: Add printing and callbacks for learning rate during training

### PR DESCRIPTION
**Pull request by madroidmaq**
Optimized the TrainingCallback definition to put some information during training into the dict, increasing the flexibility of subsequent extensions. 

Also, the corresponding learning rate information is printed during training. Here's part of the printout from when I was training:

```
Training
Starting training..., iters: 800
Iter 1: Val loss 4.847, Val took 31.577s
Iter 10: Train loss 4.461, Learning Rate 1.000e-05, It/sec 0.608, Tokens/sec 677.748, Trained Tokens 11152,
Iter 20: Train loss 3.146, Learning Rate 1.000e-05, It/sec 0.529, Tokens/sec 594.284, Trained Tokens 22383,
Iter 30: Train loss 2.157, Learning Rate 1.000e-05, It/sec 0.495, Tokens/sec 544.745, Trained Tokens 33387,
Iter 40: Train loss 1.758, Learning Rate 1.000e-05, It/sec 0.472, Tokens/sec 525.752, Trained Tokens 44534,
Iter 50: Train loss 1.471, Learning Rate 1.000e-05, It/sec 0.509, Tokens/sec 565.160, Trained Tokens 55634,
Iter 60: Train loss 1.367, Learning Rate 1.000e-05, It/sec 0.521, Tokens/sec 589.875, Trained Tokens 66965,
Iter 70: Train loss 1.230, Learning Rate 1.000e-05, It/sec 0.505, Tokens/sec 567.251, Trained Tokens 78189,
Iter 80: Train loss 1.191, Learning Rate 1.000e-05, It/sec 0.407, Tokens/sec 460.756, Trained Tokens 89497,
Iter 90: Train loss 1.073, Learning Rate 1.000e-05, It/sec 0.510, Tokens/sec 558.328, Trained Tokens 100445,
Iter 100: Train loss 1.045, Learning Rate 1.000e-05, It/sec 0.484, Tokens/sec 535.360, Trained Tokens 111513,
```

Given that the learning rate schedulers setting is supported in the `mlx` [0.3.0](https://github.com/ml-explore/mlx/releases/tag/v0.3.0) version, I added the `optim.cosine_decay ` policy to try it out, and the output is as follows The output is as follows:

```
Training
Starting training..., iters: 800
Iter 1: Val loss 4.736, Val took 31.531s
Iter 10: Train loss 3.670, Learning Rate 2.999e-05, It/sec 0.614, Tokens/sec 684.757, Trained Tokens 11152
Iter 20: Train loss 1.841, Learning Rate 2.996e-05, It/sec 0.518, Tokens/sec 582.276, Trained Tokens 22383
Iter 30: Train loss 1.308, Learning Rate 2.990e-05, It/sec 0.493, Tokens/sec 542.688, Trained Tokens 33387
Iter 40: Train loss 1.153, Learning Rate 2.982e-05, It/sec 0.483, Tokens/sec 538.068, Trained Tokens 44534
Iter 50: Train loss 1.000, Learning Rate 2.972e-05, It/sec 0.487, Tokens/sec 540.478, Trained Tokens 55634
Iter 60: Train loss 0.991, Learning Rate 2.960e-05, It/sec 0.489, Tokens/sec 554.068, Trained Tokens 66965
Iter 70: Train loss 0.921, Learning Rate 2.945e-05, It/sec 0.441, Tokens/sec 494.752, Trained Tokens 78189
Iter 80: Train loss 0.907, Learning Rate 2.928e-05, It/sec 0.322, Tokens/sec 363.570, Trained Tokens 89497
Iter 90: Train loss 0.828, Learning Rate 2.909e-05, It/sec 0.411, Tokens/sec 449.679, Trained Tokens 100445
Iter 100: Train loss 0.815, Learning Rate 2.888e-05, It/sec 0.411, Tokens/sec 454.525, Trained Tokens 111513
```

The display effect of the learning rates corresponding to the above two sets of training on wandb is roughly as follows:

image

In addition, I don't know whether it is appropriate to add `optim.cosine_decay` in the Lora.py script, so I did not submit this part of the content. I can add this part of the content if necessary. Adjustments are as follows:

```diff
+ scheduler = optim.cosine_decay(args.learning_rate, args.iters)
+ opt = optim.AdamW(learning_rate=scheduler)
- opt = optim.AdamW(learning_rate=args.learning_rate)
```

### CODE REVIEW

**Comment by madroidmaq** [comment]
Here is a simple implementation for uploading to wandb:

```py
class WandbCallback(TrainingCallback):

    def on_train_loss_report(self, train_info):
        try:
            wandb.log(train_info, step=train_info["iteration"])
        except Exception as e:
            print(f"logging to wandb failed: {e}")

    def on_val_loss_report(self, val_info):
        try:
            wandb.log(val_info, step=val_info["iteration"])
        except Exception as e:
            print(f"logging to wandb failed: {e}")
```

The corresponding wandb display looks like this:

image

**Comment by ivanfioravanti** [comment]
Great @madroidmaq I did the same locally during test 🔝

**Review by awni — APPROVED** [review]
Looks good, thanks!

**Comment by awni** [comment]
Very cool, thanks! Just curious, do you find the cosine schedule to give better results.

I think our training config needs a bit of a refactor to add support for more options. I would wait until after that to include the learning rate schedule if it's useful.

**Comment by madroidmaq** [comment]
@awni It's hard to say that cosine learning rates are always better than fixed learning rates, and when you find an optimal initial learning rate interval, cosine learning rates do outperform fixed learning rates.

I fine-tuned the test using my own private dataset, and while the cosine learning rate may be better overall when just adjusting the learning rate strategy while keeping all other parameters the same, the fixed learning rate seems to be less volatile.

Here's how my variations of different initial learning rates affected the accuracy of the final model.

image

Since I have a clear rule for determining whether the model output meets expectations, I'm not using train loos or val loss for comparison here.

**Comment by ivanfioravanti** [comment]
I confirm I have (feeling) better results using
scheduler = optim.cosine_decay(args.learning_rate, args.iters)
opt = optim.AdamW(learning_rate=scheduler)

@awni maybe we should add this as basic sample, at least people know they can play with schedulers, no?

**Comment by awni** [comment]
Yea I think that's a good idea.
